lib/web/join_tokens.go

```diff
@@ -299,7 +299,13 @@ func getJoinScript(ctx context.Context, settings scriptSettings, m nodeAPIGetter
 	}
 
 	version := proxyServers[0].GetTeleportVersion()
-	hostname, portStr, err := utils.SplitHostPort(proxyServers[0].GetPublicAddr())
+
+	publicAddr := proxyServers[0].GetPublicAddr()
+	if publicAddr == "" {
+		publicAddr = defaults.ProxyWebListenAddr().Addr
```

Comment:
doesn't `defaults.ProxyWebListenAddr()` return a constant `0.0.0.0`? Have you tested this?

It seems like `proxyServers[0].GetHostname()` would be a somewhat better fallback, but not ideal since it would probably still default to a local hostname that wouldn't be reachable by the local node.

Actually, there is precedent for falling back to the cluster name for a public proxy address, that would make the most sense in my opinion https://github.com/gravitational/teleport/blob/a4af8ae256f8382b286ba0fbfb044b67e57283a1/lib/srv/app/watcher.go#L145 https://github.com/gravitational/teleport/blob/a4af8ae256f8382b286ba0fbfb044b67e57283a1/docs/pages/reference/helm-reference/teleport-cluster.mdx?plain=1#L40

Reply:
> doesn't defaults.ProxyWebListenAddr() return a constant 0.0.0.0?

Yes, it returns `0.0.0.0:3080`. It works for the use case of running teleport locally or following the getting started on docker guide.

My reasoning is: if the user is requesting this token, they have a proxy service running. If they do and the address is empty, it defaults to `0.0.0.0:3080`, so this is what we must use when querying the install script. Does that make sense?

> Actually, there is precedent for falling back to the cluster name for a public proxy address, that would make the most sense in my opinion

That makes sense, but the cluster name doesn't have the port, so `utils.SplitHostPort(publicAddr)` will return an error.

Reply:
We can easily avoid `utils.SplitHostPort` if we want to.

Using `0.0.0.0:3080` will only work if running the proxy and the node on the same host. Maybe that will work when following the getting started guide locally, but then they will be really confused when it doesn't work if they try to run the install script on a different host.

The first thing I think we should do if we aren't able to get the true public addr is log a warning that it is not set and that the user should set it.

Then we have a number of options:

1. guess that the proxy is reachable at `0.0.0.0:3080`
2. guess that the proxy is reachable at `proxyServers[0].GetAddr()`
3. guess that the proxy is reachable at `proxyServers[0].GetHostname()+(port from GetAddr)`
4. guess that the proxy is reachable at `clusterName+(port from GetAddr)`
5. somehow extract the public hostname from the TLS cert we are serving
6. fill in the hostname on the client side
7. refuse to make a guess and return a very clear error message "proxy public_addr is not set, you must set proxy_service.public_addr to the publicly reachable address of the proxy before you can generate a node join script"

I don't think that (1) is the best option here, (2) or (3) would be strictly better IMO. (5) or (6) could be quite accurate but seem hacky and would be a good bit more work. I would personally lean toward (4) or (7).

Reply:
Thanks for that, that makes sense. I think 7 is less error-prone and points users in the right direction.

I've updated the code to log the message you suggested instead of guessing the public proxy addr.


```
2022-10-07T14:47:11-03:00 INFO [WEB]       Failed to return the node install script. error:[
ERROR REPORT:
Original Error: *errors.errorString proxy public_addr is not set, you must set proxy_service.public_addr to the publicly reachable address of the proxy before you can generate a node join script
Stack Trace:
        github.com/gravitational/teleport/lib/web/join_tokens.go:298 github.com/gravitational/teleport/lib/web.getJoinScript
        github.com/gravitational/teleport/lib/web/join_tokens.go:187 github.com/gravitational/teleport/lib/web.(*Handler).getNodeJoinScriptHandle
        github.com/gravitational/teleport/lib/httplib/httplib.go:75 github.com/gravitational/teleport/lib/httplib.MakeHandlerWithErrorWriter.func1
        github.com/julienschmidt/httprouter@v1.3.0/router.go:399 github.com/julienschmidt/httprouter.(*Router).ServeHTTP
        github.com/gravitational/teleport/lib/web/apiserver.go:235 github.com/gravitational/teleport/lib/web.(*APIHandler).ServeHTTP
        github.com/gravitational/oxy@v0.0.0-20211213172937-a1ba0900a4c9/ratelimit/tokenlimiter.go:118 github.com/gravitational/oxy/ratelimit.(*TokenLimiter).ServeHTTP
        github.com/gravitational/oxy@v0.0.0-20211213172937-a1ba0900a4c9/connlimit/connlimit.go:75 github.com/gravitational/oxy/connlimit.(*ConnLimiter).ServeHTTP
        go.opentelemetry.io/contrib/instrumentation/net/http/otelhttp@v0.36.1/handler.go:204 go.opentelemetry.io/contrib/instrumentation/net/http/otelhttp.(*Handler).ServeHTTP
        net/http/server.go:2916 net/http.serverHandler.ServeHTTP
        net/http/server.go:3523 net/http.initALPNRequest.ServeHTTP
        net/http/h2_bundle.go:5903 net/http.(*http2serverConn).runHandler
        runtime/asm_amd64.s:1571 runtime.goexit
User Message: proxy public_addr is not set, you must set proxy_service.public_addr to the publicly reachable address of the proxy before you can generate a node join script] web/join_tokens.go:189
```